### PR TITLE
OFFI-78: Support different base path for OC app

### PR DESCRIPTION
**Pull request by wAsnk**
[OFFI-78](https://lombiq.atlassian.net/browse/OFFI-78)

[OFFI-78]: https://lombiq.atlassian.net/browse/OFFI-78?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Lombiq.Tests.UI/Services/UITestExecutionSession.cs [hunk 1]
```diff
@@ -481,7 +481,7 @@ private async Task SetupAsync()
 
             _dockerConfiguration = TestConfigurationManager.GetConfiguration<DockerConfiguration>();
 
-            var resultUri = await _currentSetupSnapshotManager.RunOperationAndSnapshotIfNewAsync(async () =>
```
Comment (Piedone) [inline Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Please keep this. The setup operation needs to be able to create the URL the test should start from.

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Okay, and in what form is it used currently? How can UI tests be started on different URLs currently? I can't seem to find it.

Comment (Piedone) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: It's this feature you removed here. The URL returned by the setup operation will be the starting URL of the test.

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: I luled, obviously this is what it was using it, but it is not anymore. So let me rephrase my question. Why is it needed to use this URL to navigate to the relative URL at the end of the setup? Is there a reason for that? If so, I don't think `resultUri.PathAndQuery` is the right way of doing it.

So e.g. what you say is if you set a URL of https://example.com/path/ to start the test on, you would get the relative Url `https://example.com/path/path` at line `await _context.GoToRelativeUrlAsync(resultUri.PathAndQuery);` which is obviously not okay.
So the correct navigation step is to navigate to `/`, which means `resultUri` isn't needed.

Comment (Piedone) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Why would it be correct to always navigate to `/`? The reason we have the setup operation return a URL is precisely because that's not necessarily the case: the URL it returns is the start URL of the test. This is the URL that'll be open when control arrives at the beginning of a test's code.  The setup operation is not necessarily just filling out the setup screen, it can be anything to initialize the app to a testable state.

File: Lombiq.Tests.UI/Services/OrchardCoreInstance.cs [hunk 2]
```diff
@@ -25,6 +25,13 @@ namespace Lombiq.Tests.UI.Services;
 public class OrchardCoreConfiguration
 {
     public string SnapshotDirectoryPath { get; set; }
+
+    /// <summary>
+    /// Gets or sets the base path of the application if mapped to a different URL. For example, if the Orchard Core
+    /// application is mapped to "/cms" then this should be "/cms/". The / is important at the end, otherwise during
```
Comment (Piedone) [inline Lombiq.Tests.UI/Services/OrchardCoreInstance.cs]: You can make sure this happens with `TrimEnd('/') + "/'"`.

File: Lombiq.Tests.UI/Services/UITestExecutionSession.cs (hunk 1 above)
Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Did you understand my example?

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: > the URL it returns is the start URL of the test. This is the URL that'll be open when control arrives at the beginning of a test's code. The setup operation is not necessarily just filling out the setup screen, it can be anything to initialize the app to a testable state.

Where do you set this URL currently?

File: Lombiq.Tests.UI/SecurityScanning/SecurityScanConfiguration.cs [hunk 3]
```diff
@@ -327,7 +327,11 @@ internal async Task ApplyToPlanAsync(YamlDocument yamlDocument, UITestContext co
             //   pollPostData: ""
         }
 
+        // We can do this or .ToArray() which results IDE0305: Collection initialization can be simplified. So let's use
+        // this.
+#pragma warning disable S3878 //S3878: Remove this array creation and simply pass the elements.
         yamlDocument.AddExcludePathsRegex([.. _excludedUrlRegexPatterns]);
+#pragma warning restore S3878
```
Comment (sarahelsaig) [inline Lombiq.Tests.UI/SecurityScanning/SecurityScanConfiguration.cs]: Could you use ToArray instead? When .NET 9 comes out, it will support `params` for all `IEnumerable` types so finding these by searching for ToArray will be much easier.

File: Lombiq.Tests.UI/Services/OrchardCoreInstance.cs [hunk 4]
```diff
@@ -88,13 +95,13 @@ public async Task<Uri> StartUpAsync()
...
 
-        _reverseProxy = new TestReverseProxy(_url);
+        _reverseProxy = new TestReverseProxy(_url, _configuration.BasePath);
 
         await _reverseProxy.StartAsync();
 
         await StartOrchardAppAsync();
 
-        return new Uri(_url);
+        return new Uri(_url + _configuration.BasePath);
```
Comment (sarahelsaig) [inline Lombiq.Tests.UI/Services/OrchardCoreInstance.cs]: I think BasePath should work even if it's not prefixed with `/`. You can use [`new Uri(Uri baseUri, string relativeUri)`](https://learn.microsoft.com/en-us/dotnet/api/system.uri.-ctor?view=net-8.0#system-uri-ctor(system-uri-system-string)) instead.

**Review by sarahelsaig — CHANGES_REQUESTED** [review]

File: Lombiq.Tests.UI/Services/UITestExecutionSession.cs (hunk 1 above)
Comment (Piedone) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: This URL is the URL to start the test on. The app may launch with a different URL, i.e. the _setup_ may start with one URL, but the _test_ can start with a second one. That's what the setup operation returns.

In any case, feel free to change whatever is necessary to make the prefixed base path work, but do not remove the ability for the setup operation to return a custom start URL for a test.

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Could you show me an example where this is done? Because I don't understand how this could be done.

Comment (Piedone) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: It's in every setup operation, since otherwise you can't compile, e.g.: https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI.Samples/Helpers/SetupHelpers.cs#L46

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Okay, now I understand what you mean, but it is really weird to do this:
image
Let's say you set the basepath to start the UI tests on for `/cms` and to operate with this basepath for everything. It means the setup will be started on that page and it will return `https://localhost:port/cms/` as the Uri (homePageUri variable here), and if I return that because I want to be on that page after the setup process then it will redirect to `https://localhost:port/cms/cms`. 

So I have to return a `https://localhost:port/` Uri, because in [GoToRelativeUrlAsync](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Extensions/NavigationUITestContextExtensions.cs#L25), the [GetAbsoluteUri](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Extensions/NavigationUITestContextExtensions.cs#L71) is called to create the Uri based on the `context.Scope.BaseUri` which is set to `https://localhost:port/cms/`. So I return  `https://localhost:port/` to get `https://localhost:port/cms` at `UITestExecutionSession` `await _context.GoToRelativeUrlAsync(resultUri.PathAndQuery);`

Or is this only weird for me? I can sleep with this in my mind, it's just weird.

Comment (Piedone) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: You can try `_context.GoToRelativeUrlAsync("/" + resultUri.PathAndQuery);` if `PathAndQuery` never contains a leading slash. That's also semantically more appropriate, since the assumption there is that the URL returned from the setup operation is always relative to the app root.

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Yeah, that was my original thought. But that won't work because [GetAbsoluteUri](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Extensions/NavigationUITestContextExtensions.cs#L71) does `.TrimStart('/')` to the given Uri.

File: Lombiq.Tests.UI/BasicOrchardFeaturesTesting/MediaOperationsTestingUITestContextExtensions.cs [hunk 5]
```diff
@@ -32,7 +32,7 @@ public static class MediaOperationsTestingUITestContextExtensions
                 context.Exists(By.XPath($"//span[contains(text(), '{imageName}')]"));
 
                 await context
-                    .Get(By.CssSelector($"a[href=\"/media/{imageName}\"]").OfAnyVisibility())
```
Comment (Piedone) [inline Lombiq.Tests.UI/BasicOrchardFeaturesTesting/MediaOperationsTestingUITestContextExtensions.cs:35]: Here and below rather utilize [`UrlPrefix`](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Services/UITestContext.cs#L111).

File: Lombiq.Tests.UI/Services/UITestExecutionSession.cs (hunk 1 above)
Comment (Piedone) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: I discovered that we actually already have a mechanism for all of this, called `UrlPrefix`. So, rather do what [`SwitchCurrentTenant`](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Services/UITestContext.cs#L246) does. Everything else should work without changes.

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: image

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: I swear I did not spend several hours on this.

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Uhm this actually won't work because we also need a different base path during setup.

Comment (wAsnk) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Okay nvm, I can still do it with this, but now it just seems like another solution, and not necessarily a better solution. But I can do it this way if you say so.

Comment (Piedone) [reply Lombiq.Tests.UI/Services/UITestExecutionSession.cs]: Utilizing an existing mechanism, just also earlier, seems like the solution to go, yes.

File: Lombiq.Tests.UI/BasicOrchardFeaturesTesting/MediaOperationsTestingUITestContextExtensions.cs [hunk 6]
```diff
@@ -32,7 +32,7 @@ public static class MediaOperationsTestingUITestContextExtensions
...
+                    .Get(By.CssSelector($"a[href=\"{context.UrlPrefix}/media/{imageName}\"]").OfAnyVisibility())
```
Comment (sarahelsaig) [inline Lombiq.Tests.UI/BasicOrchardFeaturesTesting/MediaOperationsTestingUITestContextExtensions.cs]: This breaks if the image has a version query string (for cache invalidation), which will be the default behavior starting with OC 2.0. (it's already like that in [`task/system-text-json-migration`](https://github.com/Lombiq/UI-Testing-Toolbox/blob/task/system-text-json-migration/Lombiq.Tests.UI/BasicOrchardFeaturesTesting/MediaOperationsTestingUITestContextExtensions.cs#L35)) So instead of exact match you should check if the href starts with that path.
```suggestion
                    .Get(By.CssSelector($"a[href^=\"{context.UrlPrefix}/media/{imageName}\"]").OfAnyVisibility())
```

File: Lombiq.Tests.UI/BasicOrchardFeaturesTesting/MediaOperationsTestingUITestContextExtensions.cs [hunk 7]
```diff
@@ -56,7 +56,7 @@ await context
                     .ClickReliablyAsync(context);
 
                 await context
-                    .Get(By.CssSelector($"a[href=\"/media/{documentName}\"]"))
+                    .Get(By.CssSelector($"a[href=\"{context.UrlPrefix}/media/{documentName}\"]"))
```
Comment (sarahelsaig) [inline Lombiq.Tests.UI/BasicOrchardFeaturesTesting/MediaOperationsTestingUITestContextExtensions.cs]: Same.
```suggestion
                    .Get(By.CssSelector($"a[href^=\"{context.UrlPrefix}/media/{documentName}\"]"))
```

**Review by sarahelsaig — CHANGES_REQUESTED** [review]
